### PR TITLE
[MaximizeValueSemantics] Allow independent aliases

### PR DESCRIPTION
**Pull request by Groverkss**
This patch adds support to handle multiple aliases during overwriting in MaximizeValueSemantics. For example, given the following input:

```mlir
func.func @forward(%arg0: !torch.vtensor<[6,4],f32>) -> (!torch.tensor, !torch.tensor) {
  %int1 = torch.constant.int 1
  %int4 = torch.constant.int 4
  %int2 = torch.constant.int 2
  %int0 = torch.constant.int 0
  %0 = torch.tensor_static_info_cast %arg0 : !torch.vtensor<[6,4],f32> to !torch.vtensor
  %1 = torch.copy.to_tensor %0 : !torch.tensor
  %2 = torch.aten.slice.Tensor %1, %int0, %int0, %int2, %int1 : !torch.tensor, !torch.int, !torch.int, !torch.int, !torch.int -> !torch.tensor
  %3 = torch.aten.slice.Tensor %1, %int0, %int2, %int4, %int1 : !torch.tensor, !torch.int, !torch.int, !torch.int, !torch.int -> !torch.tensor
  %4 = torch.copy.to_vtensor %2 : !torch.vtensor
  %5 = torch.aten.fill.Scalar %4, %int1 : !torch.vtensor, !torch.int -> !torch.vtensor
  torch.overwrite.tensor.contents %5 overwrites %2 : !torch.vtensor, !torch.tensor
  %6 = torch.copy.to_vtensor %3 : !torch.vtensor
  %7 = torch.aten.fill.Scalar %6, %int2 : !torch.vtensor, !torch.int -> !torch.vtensor
  torch.overwrite.tensor.contents %7 overwrites %3 : !torch.vtensor, !torch.tensor
  return %2, %3 : !torch.tensor, !torch.tensor
}
```

Currently, MaximizeValueSemantics does not work on this, as it conservatively assumes that a write to `%2` is also a write to `%3`.

This patch teaches MaximizeValueSemantics when two aliases are independent of each other. This allows for lowering of lstm and lstm_cell. lstm_cell output is also added as a test in this patch.

### CODE REVIEW

**Comment by AmosLewis** [comment]
Looks like it does not solve my case.
[t5_small_torchscript_test.mlir](https://gist.github.com/AmosLewis/20895fa06e8b61ce83e9a2b9c2ee8ca6)
```
➜  t5small git:(maximizevaluesemantics) ✗ torch-mlir-opt -pass-pipeline='builtin.module(torchscript-module-to-torch-backend-pipeline{backend-legal-ops=torch.aten.flatten.using_ints,torch.aten.native_layer_norm,torch.aten.linear})' ./t5_small_torchscript_test.mlir  -mlir-print-ir-after-failure -mlir-disable-threading 
./test_torchscript.mlir:12:12: error: unsupported by backend contract: tensor with unknown rank
    %134 = torch.aten.new_zeros %arg2, %133, %int4, %int0, %cpu, %false : !torch.tensor, !torch.list<int>, !torch.int, !torch.int, !torch.Device, !torch.bool -> !torch.tensor
           ^
./test_torchscript.mlir:12:12: note: see current operation: %11 = "torch.tensor_static_info_cast"(%10) : (!torch.vtensor<[1,4],si64>) -> !torch.vtensor<*,si64>
./test_torchscript.mlir:12:12: note: this is likely due to a missing transfer function in abstract_interp_lib_gen.py
// -----// IR Dump After LowerToBackendContract Failed (torch-lower-to-backend-contract) //----- //
module attributes {torch.debug_module_name = "_lambda"} {
  func.func @forward(%arg0: !torch.vtensor<[1,15],si64>, %arg1: !torch.vtensor<[1,4],si64>) -> !torch.vtensor<*,si64> {
    %int1 = torch.constant.int 1
    %int3 = torch.constant.int 3
    %int0 = torch.constant.int 0
    %false = torch.constant.bool false
    %int4 = torch.constant.int 4
    %int-1 = torch.constant.int -1
    %none = torch.constant.none
    %int9223372036854775807 = torch.constant.int 9223372036854775807
    %cpu = torch.constant.device "cpu"
    %0 = torch.prim.ListConstruct %int1, %int4 : (!torch.int, !torch.int) -> !torch.list<int>
    %1 = torch.aten.zeros %0, %int4, %int0, %cpu, %false : !torch.list<int>, !torch.int, !torch.int, !torch.Device, !torch.bool -> !torch.vtensor<[1,4],si64>
    %2 = torch.tensor_static_info_cast %1 : !torch.vtensor<[1,4],si64> to !torch.vtensor<*,si64>
    %3 = torch.copy.to_tensor %2 : !torch.tensor<*,si64>
    %4 = torch.aten.slice.Tensor %arg1, %int1, %int0, %int-1, %int1 : !torch.vtensor<[1,4],si64>, !torch.int, !torch.int, !torch.int, !torch.int -> !torch.vtensor<[1,3],si64>
    %5 = torch.aten.clone %4, %none : !torch.vtensor<[1,3],si64>, !torch.none -> !torch.vtensor<[1,3],si64>
    %6 = torch.aten.slice.Tensor %3, %int1, %int1, %int9223372036854775807, %int1 : !torch.tensor<*,si64>, !torch.int, !torch.int, !torch.int, !torch.int -> !torch.tensor<[1,3],si64>
    %7 = torch.tensor_static_info_cast %6 : !torch.tensor<[1,3],si64> to !torch.tensor<*,si64>
    %8 = torch.prim.ListConstruct %int1, %int3 : (!torch.int, !torch.int) -> !torch.list<int>
    %9 = torch.aten.broadcast_to %5, %8 : !torch.vtensor<[1,3],si64>, !torch.list<int> -> !torch.vtensor<[1,3],si64>
    %10 = torch.tensor_static_info_cast %9 : !torch.vtensor<[1,3],si64> to !torch.vtensor<*,si64>
    torch.overwrite.tensor.contents %10 overwrites %7 : !torch.vtensor<*,si64>, !torch.tensor<*,si64>
    %11 = torch.aten.slice.Tensor %3, %int1, %int0, %int1, %int1 : !torch.tensor<*,si64>, !torch.int, !torch.int, !torch.int, !torch.int -> !torch.tensor<[1,1],si64>
    %12 = torch.aten.squeeze.dim %11, %int1 : !torch.tensor<[1,1],si64>, !torch.int -> !torch.tensor<[1],si64>
    %13 = torch.tensor_static_info_cast %12 : !torch.tensor<[1],si64> to !torch.tensor<*,si64>
    %14 = torch.copy.to_vtensor %13 : !torch.vtensor<*,si64>
    return %14 : !torch.vtensor<*,si64>
  }
}

```

File: lib/Dialect/Torch/Transforms/MaximizeValueSemantics.cpp [hunk 1]
```diff
@@ -244,6 +288,111 @@ class AbstractlyInterpretCopyToNonValueTensorOpUsersWithinABlock
...
+  static LogicalResult
+  calculateAliasInfo(Value tensor,
+                     DenseMap<Value, AliasSliceInfo> &aliasSliceInfo) {
+    Operation *op = tensor.getDefiningOp();
+
+    // If op is not a view-like op, then the tensor is an alias of itself.
+    if (!isViewLikeOp(op)) {
+      aliasSliceInfo[tensor] = AliasSliceInfo{tensor, -1, 0, -1, false};
+      return success();
+    }
```
Comment (ramiro050) [inline lib/Dialect/Torch/Transforms/MaximizeValueSemantics.cpp:302]: Once the recursion is removed, this should not be needed. I would turn it into an `assert`

File: lib/Dialect/Torch/Transforms/MaximizeValueSemantics.cpp [hunk 2]
```diff
@@ -244,6 +288,111 @@ class AbstractlyInterpretCopyToNonValueTensorOpUsersWithinABlock
...
+      if (failed(result))
+        return failure();
+
+      AliasSliceInfo &aliasInfo = aliasSliceInfo[sliceOp.getSelf()];
+
+      // We mark the slice as approximate unless we can statically determine
+      // the slicing information.
+      aliasSliceInfo[tensor] =
+          AliasSliceInfo{aliasInfo.tensor, aliasInfo.dim, aliasInfo.start,
+                         aliasInfo.length, true};
```
Comment (ramiro050) [inline lib/Dialect/Torch/Transforms/MaximizeValueSemantics.cpp:327]: nit: add `/*NameOfParam=*/true`

File: lib/Dialect/Torch/Transforms/MaximizeValueSemantics.cpp [hunk 3]
```diff
@@ -244,6 +288,111 @@ class AbstractlyInterpretCopyToNonValueTensorOpUsersWithinABlock
...
+        dim = dimInt.getValue().getSExtValue();
+      if (auto stepInt = sliceOp.getStep().getDefiningOp<ConstantIntOp>())
+        step = stepInt.getValue().getSExtValue();
+
+      // Shape of slice is not statically analyzable.
+      if (start < 0 || end < 0 || dim < 0 || step < 0)
+        return success();
+
+      // Get length of the dimension being sliced.
+      assert(end > start);
```
Comment (ramiro050) [inline lib/Dialect/Torch/Transforms/MaximizeValueSemantics.cpp:354]: This *can* happen in valid torch MLIR, so it should not be an `assert`.

File: lib/Dialect/Torch/Transforms/MaximizeValueSemantics.cpp [hunk 4]
```diff
@@ -244,6 +288,111 @@ class AbstractlyInterpretCopyToNonValueTensorOpUsersWithinABlock
     rewriteSlice(*interpretedOps, rewriter);
     return success();
   }
+
+private:
+  static LogicalResult
```
Comment (ramiro050) [inline lib/Dialect/Torch/Transforms/MaximizeValueSemantics.cpp:293]: Removing recursion should allow changing this to `void`, removing the need for the ignored return in `(void)calculateAliasInfo` above

File: test/Dialect/Torch/maximize-value-semantics.mlir [hunk 5]
```diff
@@ -261,3 +261,88 @@ func.func @viewlike$two_inputs_two_copies(%arg0: !torch.vtensor, %arg1: !torch.v
   %3 = torch.copy.to_vtensor %2 : !torch.vtensor
   return %3 : !torch.vtensor
 }
+
+// CHECK-LABEL:   func.func @multiple_aliases_independent(
```
Comment (ramiro050) [inline test/Dialect/Torch/maximize-value-semantics.mlir:265]: There are several edge cases in the code that would benefit from lit tests. For example, overlapping aliases because of `dim`, overlapping because of `start-end`, non-overlapping when one `end = -1`, use of `static_info_cast`.

**Review by ramiro050 — CHANGES_REQUESTED** [review]
Good job! A few initial comments

**Comment by Groverkss** [comment]
I'm still working on this before I need another review but I pushed some changes that make @AmosLewis 's case work and fix some bugs.

**Comment by AmosLewis** [comment]
> 

Awesome. It works well. But there is some bug if upgrading to a new llvm-project.
```
module attributes {torch.debug_module_name = "_lambda"} {
  func.func @forward(%arg0: !torch.vtensor<[1,15],si64>, %arg1: !torch.vtensor<[1,4],si64>) -> !torch.vtensor<[1],si64> {
    %int1 = torch.constant.int 1
    %int0 = torch.constant.int 0
    %false = torch.constant.bool false
    %int4 = torch.constant.int 4
    %int3 = torch.constant.int 3
    %int-1 = torch.constant.int -1
    %none = torch.constant.none
    %int9223372036854775807 = torch.constant.int 9223372036854775807
    %cpu = torch.constant.device "cpu"
    %0 = torch.prim.ListConstruct %int1, %int4 : (!torch.int, !torch.int) -> !torch.list<int>
    %1 = torch.aten.zeros %0, %int4, %int0, %cpu, %false : !torch.list<int>, !torch.int, !torch.int, !torch.Device, !torch.bool -> !torch.vtensor<[1,4],si64>
    %2 = torch.aten.slice.Tensor %arg1, %int1, %int0, %int-1, %int1 : !torch.vtensor<[1,4],si64>, !torch.int, !torch.int, !torch.int, !torch.int -> !torch.vtensor<[1,3],si64>
    %3 = torch.aten.clone %2, %none : !torch.vtensor<[1,3],si64>, !torch.none -> !torch.vtensor<[1,3],si64>
    %4 = torch.aten.slice.Tensor %1, %int1, %int1, %int9223372036854775807, %int1 : !torch.vtensor<[1,4],si64>, !torch.int, !torch.int, !torch.int, !torch.int -> !torch.vtensor<[1,3],si64>
    %5 = torch.prim.ListConstruct %int1, %int3 : (!torch.int, !torch.int) -> !torch.list<int>
    %6 = torch.aten.broadcast_to %3, %5 : !torch.vtensor<[1,3],si64>, !torch.list<int> -> !torch.vtensor<[1,3],si64>
    %7 = torch.aten.slice.Tensor %1, %int1, %int0, %int1, %int1 : !torch.vtensor<[1,4],si64>, !torch.int, !torch.int, !torch.int, !torch.int -> !torch.vtensor<[1,1],si64>
    %8 = torch.aten.squeeze.dim %7, %int1 : !torch.vtensor<[1,1],si64>, !torch.int -> !torch.vtensor<[1],si64>
    return %8 : !torch.vtensor<[1],si64>
  }
}

```

**Comment by AmosLewis** [comment]
> I'm still working on this before I need another review but I pushed some changes that make @AmosLewis 's case work and fix some bugs.


Awesome. It works well. But there is some bug if upgrading to a new llvm-project.
```
module attributes {torch.debug_module_name = "_lambda"} {
  func.func @forward(%arg0: !torch.vtensor<[1,15],si64>, %arg1: !torch.vtensor<[1,4],si64>) -> !torch.vtensor<[1],si64> {
    %int1 = torch.constant.int 1
    %int0 = torch.constant.int 0
    %false = torch.constant.bool false
    %int4 = torch.constant.int 4
    %int3 = torch.constant.int 3
    %int-1 = torch.constant.int -1
    %none = torch.constant.none
    %int9223372036854775807 = torch.constant.int 9223372036854775807
    %cpu = torch.constant.device "cpu"
    %0 = torch.prim.ListConstruct %int1, %int4 : (!torch.int, !torch.int) -> !torch.list<int>
    %1 = torch.aten.zeros %0, %int4, %int0, %cpu, %false : !torch.list<int>, !torch.int, !torch.int, !torch.Device, !torch.bool -> !torch.vtensor<[1,4],si64>
    %2 = torch.aten.slice.Tensor %arg1, %int1, %int0, %int-1, %int1 : !torch.vtensor<[1,4],si64>, !torch.int, !torch.int, !torch.int, !torch.int -> !torch.vtensor<[1,3],si64>
    %3 = torch.aten.clone %2, %none : !torch.vtensor<[1,3],si64>, !torch.none -> !torch.vtensor<[1,3],si64>
    %4 = torch.aten.slice.Tensor %1, %int1, %int1, %int9223372036854775807, %int1 : !torch.vtensor<[1,4],si64>, !torch.int, !torch.int, !torch.int, !torch.int -> !torch.vtensor<[1,3],si64>
    %5 = torch.prim.ListConstruct %int1, %int3 : (!torch.int, !torch.int) -> !torch.list<int>
    %6 = torch.aten.broadcast_to %3, %5 : !torch.vtensor<[1,3],si64>, !torch.list<int> -> !torch.vtensor<[1,3],si64>
    %7 = torch.aten.slice.Tensor %1, %int1, %int0, %int1, %int1 : !torch.vtensor<[1,4],si64>, !torch.int, !torch.int, !torch.int, !torch.int -> !torch.vtensor<[1,1],si64>
    %8 = torch.aten.squeeze.dim %7, %int1 : !torch.vtensor<[1,1],si64>, !torch.int -> !torch.vtensor<[1],si64>
    return %8 : !torch.vtensor<[1],si64>
  }
}

```

File: test/Dialect/Torch/maximize-value-semantics.mlir [hunk 6]
```diff
@@ -261,3 +261,88 @@ func.func @viewlike$two_inputs_two_copies(%arg0: !torch.vtensor, %arg1: !torch.v
...
+// CHECK-SAME:                                   %[[ARG0:.*]]: !torch.vtensor<[2,10],f32>) -> !torch.vtensor<[2,20],f32> {
+// CHECK:       %[[SLICE1:.*]] = torch.aten.slice.Tensor %[[tensor:.*]], %[[INT1:.*]], %[[INT0:.*]], %[[INT20:.*]], %[[INT1]] : !torch.vtensor<[2,80],f32>, !torch.int, !torch.int, !torch.int, !torch.int -> !torch.vtensor<[2,20],f32>
+// CHECK:       %[[SLICE2:.*]] = torch.aten.slice.Tensor %[[tensor]], %[[INT1]], %[[INT20]], %[[INT40:.*]], %[[INT1]] : !torch.vtensor<[2,80],f32>, !torch.int, !torch.int, !torch.int, !torch.int -> !torch.vtensor<[2,20],f32>
+// CHECK:       %[[SLICE3:.*]] = torch.aten.slice.Tensor %[[tensor]], %[[INT1]], %[[INT40]], %[[INT60:.*]], %[[INT1]] : !torch.vtensor<[2,80],f32>, !torch.int, !torch.int, !torch.int, !torch.int -> !torch.vtensor<[2,20],f32>
+// CHECK:       %[[SLICE4:.*]] = torch.aten.slice.Tensor %[[tensor]], %[[INT1]], %[[INT60]], %[[INT80:.*]], %[[INT1]] : !torch.vtensor<[2,80],f32>, !torch.int, !torch.int, !torch.int, !torch.int -> !torch.vtensor<[2,20],f32>
+func.func @multiple_aliases_independent_lstm_cell(%arg0: !torch.vtensor<[2,10],f32>) -> !torch.vtensor<[2,20],f32> {
```
[the remark this answers is not in the document]
Comment (Groverkss) [reply test/Dialect/Torch/maximize-value-semantics.mlir:295]: Not really. I removed it.

**Comment by Groverkss** [comment]
This is not required with the new pt2 tracing flow.
